File: src/main/java/me/testaccount666/serversystem/moderation/MuteModeration.java
package me.testaccount666.serversystem.moderation;

import java.util.UUID;

public class MuteModeration extends AbstractModeration {
    private final boolean _isShadowMute;

    public MuteModeration(boolean isShadowMute, long issueTime, long expireTime, String reason, UUID senderUuid, UUID targetUuid) {
        super(issueTime, expireTime, reason, senderUuid, targetUuid);
        _isShadowMute = isShadowMute;
    }

    public boolean isShadowMute() {
        return _isShadowMute;
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/moderation/MuteModeration.kt
package me.testaccount666.serversystem.moderation

class MuteModeration : AbstractModeration {
    val isShadowMute: Boolean

    private constructor(b: MuteModerationBuilder<*>) : super(b) {
        isShadowMute = b.isShadowMute
    }

    abstract class MuteModerationBuilder<B : MuteModerationBuilder<B>> : AbstractModerationBuilder<MuteModeration, B>() {
        internal var isShadowMute = false

        fun isShadowMute(isShadowMute: Boolean): B {
            this.isShadowMute = isShadowMute
            return self()
        }

        abstract override fun self(): B

        abstract override fun build(): MuteModeration

        override fun toString() = "MuteModeration.MuteModerationBuilder(super=${super.toString()}, isShadowMute=$isShadowMute)"
    }

    private class MuteModerationBuilderImpl : MuteModerationBuilder<MuteModerationBuilderImpl>() {
        override fun self() = this

        override fun build() = MuteModeration(this)
    }

    companion object {
        @JvmStatic
        fun builder(): MuteModerationBuilder<*> = MuteModerationBuilderImpl()
    }
}
